heft-plugins/heft-lint-plugin/src/SarifFormatter.ts

```diff
@@ -0,0 +1,270 @@
+}
+
+interface IMessage extends TEslint.Linter.LintMessage {
+  suppressions?: ISuppressedAnalysis[];
+}
+
+interface IExtendedLintResult extends TEslint.ESLint.LintResult {
+  suppressedMessages: TEslint.ESLint.LintResult['suppressedMessages'];
+}
+
+const sarifFiles: Map<string, ISarifFile> = new Map();
+const sarifResults: ISarifRepresentation[] = [];
+
+const internalErrorId: string = 'ESL0999';
+const toolConfigurationNotifications: ISarifRepresentation[] = [];
+
+/**
+ * Converts ESLint results into a SARIF (Static Analysis Results Interchange Format) log.
+ *
+ * This function takes in a list of ESLint lint results, processes them to extract
+ * relevant information such as errors, warnings, and suppressed messages, and
+ * outputs a SARIF log which conforms to the SARIF v2.1.0 specification.
+ *
+ * @param results - An array of lint results (`IExtendedLintResult[]`) from ESLint that contains linting information,
+ *                  such as file paths, messages, and suppression details.
+ * @param options - An object (`ISerifFormatterOptions`) containing options for formatting:
+ *                  - `ignoreSuppressed`: Boolean flag to decide whether to ignore suppressed messages.
+ *                  - `eslintVersion`: Optional string to include the version of ESLint in the SARIF log.
+ * @returns The SARIF log (`ISarifLog`) containing information about the linting results in SARIF format.
+ */
+export function formatAsSARIF(results: IExtendedLintResult[], options: ISerifFormatterOptions): ISarifLog {
+  const { ignoreSuppressed, eslintVersion } = options;
+  const sarifRun: ISarifRun = {
+    tool: {
+      driver: {
+        name: 'ESLint',
+        informationUri: 'https://eslint.org',
+        rules: []
+      }
+    }
+  };
+
+  const sarifLog: ISarifLog = {
+    version: '2.1.0',
+    $schema: 'http://json.schemastore.org/sarif-2.1.0-rtm.5',
+    runs: [sarifRun]
+  };
+
+  if (typeof eslintVersion !== undefined) {
+    sarifRun.tool.driver.version = eslintVersion;
+  }
+  let executionSuccessful: boolean = true;
+
+  for (const result of results) {
+    const { filePath } = result;
+    let sarifFile: ISarifFile | undefined = sarifFiles.get(filePath);
+    if (sarifFile === undefined) {
+      const artifactIndex: number = sarifFiles.size;
+      const fileUrl: string = url.pathToFileURL(filePath).toString();
+
+      sarifFile = {
+        location: {
+          uri: fileUrl
+        }
+      };
+      sarifFiles.set(filePath, sarifFile);
+
+      const containsSuppressedMessages: boolean =
+        result.suppressedMessages && result.suppressedMessages.length > 0;
+      const messages: IMessage[] =
+        containsSuppressedMessages && !ignoreSuppressed
+          ? [...result.messages, ...result.suppressedMessages]
+          : result.messages;
+
+      if (messages.length > 0) {
+        for (const message of messages) {
+          const level: string = message.fatal || message.severity === 2 ? 'error' : 'warning';
```

Comment:
Is there an actual const that exists to compare the severity against? `2` will work, but better to compare against a known-value for the severity type

Reply:
For SARIF, the initial type severity is 0, 1, or 2 and then it is assigned a string const "error", "warning" or "note."